File: src/graph/CameraBoxPainter.js
import {parsegraph_defaultFont} from './settings';
import {timediffMs} from '../timing';
import Color from './Color';
import parsegraph_Label from './Label';
import parsegraph_GlyphPainter from './GlyphPainter';
import BlockPainter from './BlockPainter';

export default function parsegraph_CameraBoxPainter(window) {
  this._blockPainter = new BlockPainter(window);
  this._glyphPainter = new parsegraph_GlyphPainter(
      window,
      parsegraph_defaultFont(),
  );

  this._borderColor = new Color(1, 1, 1, 0.1);
  this._backgroundColor = new Color(1, 1, 1, 0.1);
  this._textColor = new Color(1, 1, 1, 1);
  this._fontSize = 24;
}

parsegraph_CameraBoxPainter.prototype.contextChanged = function(isLost) {
  if (!isLost) {
    return;
  }
  this._blockPainter.contextChanged(isLost);
  this._glyphPainter.contextChanged(isLost);
};

parsegraph_CameraBoxPainter.prototype.clear = function() {
  this._glyphPainter.clear();
  this._blockPainter.clear();
};

parsegraph_CameraBoxPainter.prototype.drawBox = function(
    name,
    rect,
    scale,
    mouseX,
    mouseY,
    when,
) {
  const painter = this._blockPainter;

  const now = new Date();
  const diff = timediffMs(when, now);
  const zc = new Color(0, 0, 0, 0);
  let interp = 1;
  const fadeDelay = 500;
  const fadeLength = 1000;
  if (diff > fadeDelay) {
    interp = 1 - (diff - fadeDelay) / fadeLength;
  }
  this._borderColor.setA(0.1 * interp);
  this._backgroundColor.setA(0.1 * interp);
  painter.setBorderColor(this._borderColor);
  painter.setBackgroundColor(this._backgroundColor);
  this._glyphPainter.color().setA(interp);
  this._glyphPainter.backgroundColor().setA(interp);

  painter.drawBlock(
      rect.x(),
      rect.y(),
      rect.width(),
      rect.height(),
      0.01,
      0.1,
      scale,
  );
  const font = this._glyphPainter.font();
  const label = new parsegraph_Label(font);
  label.setText(name);
  const lw = (label.width() * (this._fontSize / font.fontSize())) / scale;
  const lh = (label.height() * (this._fontSize / font.fontSize())) / scale;

  if (mouseX === undefined) {
    mouseX = rect.width() / 2 - lw;
  }
  if (mouseY === undefined) {
    mouseY = 0;
  }
  if (mouseX < 0) {
    mouseX = 0;
  }
  mouseX = Math.min(mouseX, rect.width());
  mouseY = Math.min(mouseY, rect.height());
  if (mouseY < 0) {
    mouseY = 0;
  }

  label.paint(
      this._glyphPainter,
      rect.x() - lw / 2 - rect.width() / 2 + mouseX / scale,
      rect.y() - lh / 2 - rect.height() / 2 + mouseY / scale,
      this._fontSize / font.fontSize() / scale,
  );

  return interp > 0;
};

parsegraph_CameraBoxPainter.prototype.render = function(world, scale) {
  this._blockPainter.render(world, scale);
  this._glyphPainter.render(world, scale);
};

File: src/graph/CameraBoxPainter.c
#include "CameraBoxPainter.h"
#include "BlockPainter.h"
#include "Surface.h"
#include "GlyphAtlas.h"
#include "Color.h"
#include "Rect.h"
#include "Label.h"

parsegraph_CameraBoxPainter* parsegraph_CameraBoxPainter_new(parsegraph_Surface* surface, parsegraph_GlyphAtlas* glyphAtlas, apr_hash_t* shaders)
{
    apr_pool_t* pool = surface->pool;
    parsegraph_CameraBoxPainter* painter = apr_palloc(pool, sizeof(*painter));
    painter->_pool = pool;
    painter->_blockPainter = parsegraph_BlockPainter_new(pool, shaders);
    painter->_glyphPainter = parsegraph_GlyphPainter_new(pool, glyphAtlas, shaders);

    painter->_glyphAtlas = glyphAtlas;
    parsegraph_Color_SetRGBA(painter->_borderColor, 1, 1, 1, 0.1);
    parsegraph_Color_SetRGBA(painter->_backgroundColor, 1, 1, 1, 0.1);
    parsegraph_Color_SetRGBA(painter->_textColor, 1, 1, 1, 1);
    painter->_fontSize = 24;

    return painter;
}

void parsegraph_CameraBoxPainter_destroy(parsegraph_CameraBoxPainter* painter)
{
    parsegraph_GlyphPainter_destroy(painter->_glyphPainter);
    parsegraph_BlockPainter_destroy(painter->_blockPainter);
}

void parsegraph_CameraBoxPainter_clear(parsegraph_CameraBoxPainter* painter)
{
    parsegraph_GlyphPainter_clear(painter->_glyphPainter);
    parsegraph_BlockPainter_clear(painter->_blockPainter);
}

void parsegraph_CameraBoxPainter_drawBox(parsegraph_CameraBoxPainter* painter, const UChar* name, int len, float* rect, float scale)
{
    parsegraph_BlockPainter* blockPainter = painter->_blockPainter;
    parsegraph_BlockPainter_setBorderColor(blockPainter, painter->_borderColor);
    parsegraph_BlockPainter_setBackgroundColor(blockPainter, painter->_backgroundColor);
    parsegraph_BlockPainter_drawBlock(blockPainter,
        parsegraph_Rect_x(rect),
        parsegraph_Rect_y(rect),
        parsegraph_Rect_width(rect),
        parsegraph_Rect_height(rect),
        0.01,
        .1,
        scale
    );

    parsegraph_Label* label = parsegraph_Label_new(painter->_pool, painter->_glyphAtlas);
    parsegraph_Label_setText(label, name, len);
    float lw = parsegraph_Label_width(label)*(painter->_fontSize/parsegraph_GlyphAtlas_fontSize(painter->_glyphAtlas))/scale;

    parsegraph_Label_paint(label, painter->_glyphPainter,
        parsegraph_Rect_x(rect) - lw/2,
        parsegraph_Rect_y(rect) - parsegraph_Rect_height(rect)/2,
        (painter->_fontSize/parsegraph_GlyphAtlas_fontSize(painter->_glyphAtlas))/scale
    );

    parsegraph_Label_destroy(label);
}

void parsegraph_CameraBoxPainter_render(parsegraph_CameraBoxPainter* painter, float* world, float scale)
{
    parsegraph_BlockPainter_render(painter->_blockPainter, world, scale);
    parsegraph_GlyphPainter_render(painter->_glyphPainter, world, scale);
}
